File: pathfind_lua/PathFinder.lua
-- PathFinder.lua
-- Port of PathFinder.cpp to Lua 5.1
-- Loads .mmap/.mmtile navmesh tiles and .map terrain files.

local NavMeshModule  = require("NavMesh")
local NavMeshQuery   = require("NavMeshQuery")
local TerrainMap     = require("TerrainMap")
local loaderModule = (type(_G) == "table" and rawget(_G, "MmapLoaderModule")) or nil
if type(loaderModule) ~= "string" or loaderModule == "" then
    loaderModule = "MmapFileLoader"
end
local MmapLoader     = require(loaderModule)

local TERRAIN_INVALID_HEIGHT = TerrainMap.TERRAIN_INVALID_HEIGHT

-- Set PathFinderDebug = false before requiring this module to suppress log output.
local function _log(msg)
    local qhs = rawget(_G, "QHS")
    if type(qhs) == "table" and type(qhs.chat) == "function" then
        qhs.chat(msg)
    elseif type(io) == "table" and io.stderr and type(io.stderr.write) == "function" then
        io.stderr:write(msg .. "\n")
    elseif type(DEFAULT_CHAT_FRAME) == "table" and type(DEFAULT_CHAT_FRAME.AddMessage) == "function" then
        DEFAULT_CHAT_FRAME:AddMessage(msg)
    elseif type(print) == "function" then
        print(msg)
    end
end
local function dbg(fmt, ...)
    if PathFinderDebug == false then return end
    _log(string.format("[pathfind] " .. fmt, ...))
end
-- info() always logs regardless of PathFinderDebug.
local function info(fmt, ...)
    _log(string.format("[pathfind] " .. fmt, ...))
end

local M = {}

local BLOCK_SIZE  = 533.33333
local TILE_ORIGIN = 32
local MAX_POLYS   = 4096
local WORLD_MAX   = 17066.666

-- ---------------------------------------------------------------------------
-- Coordinate conversion helpers
-- ---------------------------------------------------------------------------

-- WoW (x, y, z) -> Recast (y, z, x)
local function toRecast(x, y, z)
    return {y, z, x}
end

-- worldToTile: compute tile coords from WoW (x,y)
-- Returns tx, ty or nil on failure
local function worldToTile(x, y)
    local ftx = TILE_ORIGIN - x / BLOCK_SIZE
    local fty = TILE_ORIGIN - y / BLOCK_SIZE
    if ftx < 0 or ftx >= 64 or fty < 0 or fty >= 64 then
        dbg("ERROR: coords (%.2f, %.2f) out of world bounds", x, y)
        return nil
    end
    return math.floor(ftx), math.floor(fty)
end

local function packTileID(tx, ty)
    return tx * 65536 + ty
end

-- ---------------------------------------------------------------------------
-- PathFinder constructor
-- ---------------------------------------------------------------------------

function M.new(dataDir, mapId)
    local pf = {
        _dataDir         = dataDir,
        _mapId           = mapId,
        _navMesh         = nil,
        _navQuery        = nil,
        _loadedNavTiles  = {},   -- set: tileID -> true
        _terrainTiles    = {},   -- map: tileID -> TerrainMap
        _initialized     = false,
        -- Pre-allocated per-query buffers (avoids alloc on every computeDistance call)
        _startPos        = {0,0,0},
        _endPos          = {0,0,0},
        _endPosAdj       = {0,0,0},
        _extents         = {2.0, 4.0, 2.0},
    }

    -- Load .mmap params file
    -- dtNavMeshParams: orig[3] + tileWidth + tileHeight + maxTiles + maxPolys = 7*4 = 28 bytes
    local paramsData, mmapPath = MmapLoader.loadNavMeshParams(dataDir, mapId)
    dbg("loading navmesh params: %s", mmapPath)
    if not paramsData then
        dbg("ERROR: could not read %s", mmapPath)
        return pf
    end

    -- Create NavMesh
    local navmesh = NavMeshModule.new(paramsData)
    pf._navMesh = navmesh

    -- Create NavMeshQuery
    pf._navQuery = NavMeshQuery.new(navmesh, 65535)
    pf._filter   = NavMeshQuery.newFilter()

    dbg("navmesh ready")
    pf._initialized = true

    -- loadNavTile(tx, ty)
    function pf:loadNavTile(tx, ty)
        local id = packTileID(tx, ty)
        if self._loadedNavTiles[id] then return true end

        local data, path = MmapLoader.loadNavTile(self._dataDir, self._mapId, tx, ty)
        dbg("loading nav tile (%d,%d): %s", tx, ty, path)
        if not data then
            dbg("WARNING: nav tile (%d,%d) not found or too short — skipping", tx, ty)
            self._loadedNavTiles[id] = true  -- mark as tried so we don't retry on every query
            return false
        end

        local ok, err = self._navMesh:addTile(data)
        if not ok then
            -- Always log addTile failures; they indicate data or parse problems.
            local msg = string.format("[pathfind] ERROR: addTile(%d,%d) failed: %s", tx, ty, tostring(err))
            local qhs = rawget(_G, "QHS")
            if type(qhs) == "table" and type(qhs.chat) == "function" then
                qhs.chat(msg)
            elseif type(DEFAULT_CHAT_FRAME) == "table" and type(DEFAULT_CHAT_FRAME.AddMessage) == "function" then
                DEFAULT_CHAT_FRAME:AddMessage(msg)
            elseif type(print) == "function" then
                print(msg)
            end
            return false
        end

        dbg("nav tile (%d,%d) loaded OK (%d bytes)", tx, ty, #data)
        self._loadedNavTiles[id] = true
        return true
    end

    -- ensureNavTilesLoaded(x1, y1, x2, y2)
    -- Loads all nav tiles covering the bounding box of (x1,y1)-(x2,y2), plus TILE_PADDING
    -- tiles of margin on each side.  A* paths can leave the straight-line bounding box when
    -- routing around terrain, so the extra margin prevents "tile not loaded" failures.
    local TILE_PADDING = 1
    function pf:ensureNavTilesLoaded(x1, y1, x2, y2)
        local tx1, ty1 = worldToTile(x1, y1)
        local tx2, ty2 = worldToTile(x2, y2)
        if not tx1 or not tx2 then return end

        -- Lua analyzer may infer nil; guard is above, values are integers here.
        local txMin = math.max(0,  math.min(tx1, tx2) - TILE_PADDING)
        local txMax = math.min(63, math.max(tx1, tx2) + TILE_PADDING)
        local tyMin = math.max(0,  math.min(ty1, ty2) - TILE_PADDING)
        local tyMax = math.min(63, math.max(ty1, ty2) + TILE_PADDING)

        dbg("tile range x:[%d,%d] y:[%d,%d]", txMin or 0, txMax or 0, tyMin or 0, tyMax or 0)

        for tx = txMin, txMax do
            for ty = tyMin, tyMax do
                self:loadNavTile(tx, ty)
            end
        end
    end

    -- _pendingTiles(x1, y1, x2, y2) -> array of {tx, ty} not yet loaded
    function pf:_pendingTiles(x1, y1, x2, y2)
        local tx1, ty1 = worldToTile(x1, y1)
        local tx2, ty2 = worldToTile(x2, y2)
        if not tx1 or not tx2 then return {} end

        local txMin = math.max(0,  math.min(tx1, tx2) - TILE_PADDING)
        local txMax = math.min(63, math.max(tx1, tx2) + TILE_PADDING)
        local tyMin = math.max(0,  math.min(ty1, ty2) - TILE_PADDING)
        local tyMax = math.min(63, math.max(ty1, ty2) + TILE_PADDING)

        local pending = {}
        for tx = txMin, txMax do
            for ty = tyMin, tyMax do
                if not self._loadedNavTiles[packTileID(tx, ty)] then
                    pending[#pending + 1] = {tx, ty}
                end
            end
        end
        return pending
    end

    -- _scheduleAsync(workFn, onDone)
    -- Calls workFn() once per C_Timer.After(0) tick until it returns true, then calls onDone().
    -- Falls back to synchronous execution if C_Timer is unavailable.
    local function _scheduleAsync(workFn, onDone)
        local C_Timer = rawget(_G, "C_Timer")
        if type(C_Timer) ~= "table" or type(C_Timer.After) ~= "function" then
            -- No timer available (external/test env): run synchronously.
            while not workFn() do end
            onDone()
            return
        end
        local function tick()
            if workFn() then
                onDone()
            else
                C_Timer.After(0, tick)
            end
        end
        C_Timer.After(0, tick)
    end

    -- computePathAsync(x1, y1, x2, y2, callback)
    -- Loads required nav tiles one per frame tick, then runs the pathfind query.
    -- callback(waypoints, total_dist) on success, callback(nil, -1) on failure.
    function pf:computePathAsync(x1, y1, x2, y2, callback)
        dbg("computePathAsync (%.4f, %.4f) -> (%.4f, %.4f)", x1, y1, x2, y2)
        local pending = self:_pendingTiles(x1, y1, x2, y2)
        local i = 0
        _scheduleAsync(function()
            i = i + 1
            if i <= #pending then
                local t = pending[i]
                self:loadNavTile(t[1], t[2])
                return false  -- more tiles to load
            end
            return true  -- all tiles loaded
        end, function()
            info("tiles loaded (%d total), computing path...", #pending)
            local straightResult, nstraight, total = self:_run_query(x1, y1, x2, y2)
            if not straightResult then
                callback(nil, -1.0)
                return
            end
            local waypoints = {}
            for k = 1, nstraight do
                local pt = straightResult[k]
                waypoints[k] = { pt[3], pt[1] }
            end
            callback(waypoints, total)
        end)
    end

    -- computeDistanceAsync(x1, y1, x2, y2, callback)
    -- Loads required nav tiles one per frame tick, then computes path distance.
    -- callback(distance) where distance >= 0 on success, -1 on failure.
    function pf:computeDistanceAsync(x1, y1, x2, y2, callback)
        dbg("computeDistanceAsync (%.4f, %.4f) -> (%.4f, %.4f)", x1, y1, x2, y2)
        local pending = self:_pendingTiles(x1, y1, x2, y2)
        local i = 0
        _scheduleAsync(function()
            i = i + 1
            if i <= #pending then
                local t = pending[i]
                self:loadNavTile(t[1], t[2])
                return false
            end
            return true
        end, function()
            info("tiles loaded (%d total), computing distance...", #pending)
            local _, _, total = self:_run_query(x1, y1, x2, y2)
            callback(total)
        end)
    end

    -- getTerrainHeight(x, y) -> height
    function pf:getTerrainHeight(x, y)
        local tx, ty = worldToTile(x, y)
        if not tx then return TERRAIN_INVALID_HEIGHT end

        local id = packTileID(tx, ty)
        if not self._terrainTiles[id] then
            local data, path = MmapLoader.loadTerrainTile(self._dataDir, self._mapId, tx, ty)
            dbg("loading terrain tile (%d,%d): %s", tx, ty, path)

            local tm = TerrainMap.new(data)
            self._terrainTiles[id] = tm

            local ok = tm:isLoaded()
            dbg("terrain tile (%d,%d) %s", tx, ty, ok and "loaded OK" or "FAILED (no .map file?)")
        end

        local h = self._terrainTiles[id]:getHeight(x, y)
        dbg("terrain height at (%.2f, %.2f) = %.4f", x, y, h)
        return h
    end

    -- _run_query(x1, y1, x2, y2) -> straightResult, nstraight, total  or  nil, 0, -1
    -- Shared body used by computeDistance and computePath.
    function pf:_run_query(x1, y1, x2, y2)
        if math.abs(x1) > WORLD_MAX or math.abs(y1) > WORLD_MAX or
           math.abs(x2) > WORLD_MAX or math.abs(y2) > WORLD_MAX then
            dbg("ERROR: coordinates exceed world bounds")
            return nil, 0, -1.0
        end

        if not self._initialized then
            dbg("ERROR: not initialized")
            return nil, 0, -1.0
        end

        self:ensureNavTilesLoaded(x1, y1, x2, y2)

        local z1 = self:getTerrainHeight(x1, y1)
        local z2 = self:getTerrainHeight(x2, y2)

        if z1 == TERRAIN_INVALID_HEIGHT or z2 == TERRAIN_INVALID_HEIGHT then
            dbg("ERROR: terrain height lookup failed (z1=%.2f z2=%.2f)", z1, z2)
            return nil, 0, -1.0
        end

        -- Fill pre-allocated recast buffers (avoids 2 table allocs per call)
        local startPos = self._startPos
        startPos[1]=y1; startPos[2]=z1; startPos[3]=x1
        local endPos = self._endPos
        endPos[1]=y2; endPos[2]=z2; endPos[3]=x2
        dbg("recast start=(%.2f,%.2f,%.2f) end=(%.2f,%.2f,%.2f)",
            startPos[1], startPos[2], startPos[3],
            endPos[1],   endPos[2],   endPos[3])

        local extents = self._extents   -- pre-allocated {2.0, 4.0, 2.0}
        local filter  = self._filter

        local startRef, startNearPt = self._navQuery:findNearestPoly(startPos, extents, filter)
        local endRef,   endNearPt   = self._navQuery:findNearestPoly(endPos,   extents, filter)

        dbg("startRef=%.0f endRef=%.0f", startRef or 0, endRef or 0)

        if not startRef or startRef == 0 or not endRef or endRef == 0 then
            dbg("ERROR: could not find nearest poly")
            return nil, 0, -1.0
        end

        local path, npolys, status = self._navQuery:findPath(
            startRef, endRef, startPos, endPos, filter, MAX_POLYS, info)
        info("findPath done: %d polys status=0x%x%s", npolys, status or 0,
            (status and (status % 256) >= 64) and " (PARTIAL)" or "")

        if npolys == 0 then return nil, 0, -1.0 end

        -- Clamp end to last reachable poly if incomplete (pre-alloc buffer)
        local endPosAdj = self._endPosAdj
        endPosAdj[1]=endPos[1]; endPosAdj[2]=endPos[2]; endPosAdj[3]=endPos[3]
        if path[npolys] ~= endRef then
            dbg("path incomplete — clamping to last reachable poly")
            local cp, _ = self._navQuery:closestPointOnPoly(path[npolys], endPos)
            if cp then
                endPosAdj[1] = cp[1]; endPosAdj[2] = cp[2]; endPosAdj[3] = cp[3]
            end
        end

        local straightResult, nstraight, _ = self._navQuery:findStraightPath(
            startPos, endPosAdj, path, npolys, MAX_POLYS, 0)

        dbg("findStraightPath: %d points", nstraight)

        if nstraight < 2 then return nil, 0, -1.0 end

        local total = 0.0
        for k = 1, nstraight - 1 do
            local a = straightResult[k]
            local b = straightResult[k+1]
            local dx = b[1]-a[1]; local dy = b[2]-a[2]; local dz = b[3]-a[3]
            total = total + math.sqrt(dx*dx + dy*dy + dz*dz)
        end

        dbg("total distance: %.4f", total)
        return straightResult, nstraight, total
    end

    -- computeDistance(x1, y1, x2, y2) -> distance or -1
    function pf:computeDistance(x1, y1, x2, y2)
        dbg("computeDistance (%.4f, %.4f) -> (%.4f, %.4f)", x1, y1, x2, y2)
        local _, _, total = self:_run_query(x1, y1, x2, y2)
        return total
    end

    -- computePath(x1, y1, x2, y2) -> waypoints, total_dist  or  nil, -1
    -- waypoints is an array of {wx, wy} in WoW world-yard coordinates.
    -- Recast stores points as {wow_y, wow_z, wow_x}, so wx=[3], wy=[1].
    function pf:computePath(x1, y1, x2, y2)
        dbg("computePath (%.4f, %.4f) -> (%.4f, %.4f)", x1, y1, x2, y2)
        local straightResult, nstraight, total = self:_run_query(x1, y1, x2, y2)
        if not straightResult then
            return nil, -1.0
        end

        local waypoints = {}
        for k = 1, nstraight do
            local pt = straightResult[k]   -- {wow_y, wow_z, wow_x}
            waypoints[k] = { pt[3], pt[1] }  -- {wx, wy}
        end

        return waypoints, total
    end

    function pf:isValid()
        return self._initialized
    end

    return pf
end

if package and package.loaded then
    package.loaded["PathFinder"] = M
end
return M

File: PathFinder.cpp
#include "PathFinder.h"

#include "DetourNavMesh.h"
#include "DetourNavMeshQuery.h"
#include "DetourCommon.h"

#include <cstdio>
#include <cmath>
#include <algorithm>

// Must match the header written by MoveMapGen into each .mmtile file.
#define MMAP_MAGIC   0x4d4d4150u  // 'MMAP'
#define MMAP_VERSION 8u

// Valid WoW world coordinate range (yards from map center)
static constexpr float WORLD_MAX = 17066.666f;

struct MmapTileHeader
{
    uint32_t mmapMagic;
    uint32_t dtVersion;
    uint32_t mmapVersion;
    uint32_t size;
    uint32_t usesLiquids;
};

static inline uint32_t packTileID(uint32_t tx, uint32_t ty)
{
    return (tx << 16) | ty;
}

// ---------------------------------------------------------------------------

PathFinder::PathFinder(const std::string& dataDir, uint32_t mapId)
    : m_dataDir(dataDir)
    , m_mapId(mapId)
    , m_navMesh(nullptr)
    , m_navQuery(nullptr)
    , m_initialized(false)
{
    char path[512];
    snprintf(path, sizeof(path), "%s/mmaps/%03u.mmap", dataDir.c_str(), mapId);
    fprintf(stderr, "[pathfind] loading navmesh params: %s\n", path);

    FILE* fp = fopen(path, "rb");
    if (!fp)
    {
        fprintf(stderr, "[pathfind] ERROR: could not open %s\n", path);
        return;
    }

    dtNavMeshParams params;
    fread(&params, sizeof(params), 1, fp);
    fclose(fp);

    fprintf(stderr, "[pathfind] navmesh params: orig=(%.2f,%.2f,%.2f) tileW=%.2f tileH=%.2f maxTiles=%d maxPolys=%d\n",
            params.orig[0], params.orig[1], params.orig[2],
            params.tileWidth, params.tileHeight,
            params.maxTiles, params.maxPolys);

    m_navMesh = dtAllocNavMesh();
    if (!m_navMesh || dtStatusFailed(m_navMesh->init(&params)))
    {
        fprintf(stderr, "[pathfind] ERROR: dtNavMesh::init failed\n");
        return;
    }

    m_navQuery = dtAllocNavMeshQuery();
    if (!m_navQuery || dtStatusFailed(m_navQuery->init(m_navMesh, 65535)))
    {
        fprintf(stderr, "[pathfind] ERROR: dtNavMeshQuery::init failed\n");
        return;
    }

    fprintf(stderr, "[pathfind] navmesh ready\n");
    m_initialized = true;
}

PathFinder::~PathFinder()
{
    if (m_navQuery) dtFreeNavMeshQuery(m_navQuery);
    if (m_navMesh)  dtFreeNavMesh(m_navMesh);
}

// ---------------------------------------------------------------------------

// WoW world space (x, y, z) -> Recast space (y, z, x)
// Matches the conversion in cmangos PathFinder::BuildPolyPath (PathFinder.cpp:411).
void PathFinder::toRecast(float x, float y, float z, float* out)
{
    out[0] = y;
    out[1] = z;
    out[2] = x;
}

// Tile coords are computed from WoW x/y directly, same formula as the server's
// MoveMap loader and RecastDemoMod's GetGridCoord.
bool PathFinder::worldToTile(float x, float y, uint32_t& tx, uint32_t& ty) const
{
    float ftx = TILE_ORIGIN - x / BLOCK_SIZE;
    float fty = TILE_ORIGIN - y / BLOCK_SIZE;
    if (ftx < 0.0f || ftx >= 64.0f || fty < 0.0f || fty >= 64.0f)
    {
        fprintf(stderr, "[pathfind] ERROR: coordinates (%.2f, %.2f) out of world bounds "
                        "(tile %.2f, %.2f — expected [0,64))\n", x, y, ftx, fty);
        return false;
    }
    tx = static_cast<uint32_t>(ftx);
    ty = static_cast<uint32_t>(fty);
    return true;
}

bool PathFinder::loadNavTile(uint32_t tx, uint32_t ty)
{
    uint32_t id = packTileID(tx, ty);
    if (m_loadedNavTiles.count(id))
        return true;

    char path[512];
    snprintf(path, sizeof(path), "%s/mmaps/%03u%02u%02u.mmtile",
             m_dataDir.c_str(), m_mapId, tx, ty);
    fprintf(stderr, "[pathfind] loading nav tile (%u,%u): %s\n", tx, ty, path);

    FILE* fp = fopen(path, "rb");
    if (!fp)
    {
        fprintf(stderr, "[pathfind] WARNING: nav tile not found\n");
        return false;
    }

    MmapTileHeader header;
    fread(&header, sizeof(header), 1, fp);

    if (header.mmapMagic != MMAP_MAGIC || header.mmapVersion != MMAP_VERSION)
    {
        fprintf(stderr, "[pathfind] ERROR: bad mmtile header (magic=0x%08x ver=%u)\n",
                header.mmapMagic, header.mmapVersion);
        fclose(fp);
        return false;
    }

    unsigned char* data = static_cast<unsigned char*>(dtAlloc(header.size, DT_ALLOC_PERM));
    if (!data)
    {
        fclose(fp);
        return false;
    }

    fread(data, header.size, 1, fp);
    fclose(fp);

    dtTileRef tileRef = 0;
    if (dtStatusFailed(m_navMesh->addTile(data, header.size, DT_TILE_FREE_DATA, 0, &tileRef)))
    {
        fprintf(stderr, "[pathfind] ERROR: addTile failed\n");
        dtFree(data);
        return false;
    }

    fprintf(stderr, "[pathfind] nav tile (%u,%u) loaded OK (%u bytes)\n", tx, ty, header.size);
    m_loadedNavTiles.insert(id);
    return true;
}

void PathFinder::ensureNavTilesLoaded(float x1, float y1, float x2, float y2)
{
    uint32_t tx1, ty1, tx2, ty2;
    if (!worldToTile(x1, y1, tx1, ty1)) return;
    if (!worldToTile(x2, y2, tx2, ty2)) return;

    uint32_t txMin = std::min(tx1, tx2);
    uint32_t txMax = std::max(tx1, tx2);
    uint32_t tyMin = std::min(ty1, ty2);
    uint32_t tyMax = std::max(ty1, ty2);

    fprintf(stderr, "[pathfind] tile range x:[%u,%u] y:[%u,%u]\n", txMin, txMax, tyMin, tyMax);

    for (uint32_t tx = txMin; tx <= txMax; ++tx)
        for (uint32_t ty = tyMin; ty <= tyMax; ++ty)
            loadNavTile(tx, ty);
}

float PathFinder::getTerrainHeight(float x, float y)
{
    uint32_t tx, ty;
    if (!worldToTile(x, y, tx, ty))
        return TERRAIN_INVALID_HEIGHT;

    uint32_t id = packTileID(tx, ty);

    auto it = m_terrainTiles.find(id);
    if (it == m_terrainTiles.end())
    {
        TerrainMap& tile = m_terrainTiles[id];

        char path[512];
        snprintf(path, sizeof(path), "%s/maps/%03u%02u%02u.map",
                 m_dataDir.c_str(), m_mapId, tx, ty);
        fprintf(stderr, "[pathfind] loading terrain tile (%u,%u): %s\n", tx, ty, path);

        bool ok = tile.load(path);
        fprintf(stderr, "[pathfind] terrain tile (%u,%u) %s\n", tx, ty, ok ? "loaded OK" : "FAILED (no .map file?)");

        float h = tile.getHeight(x, y);
        fprintf(stderr, "[pathfind] terrain height at (%.2f, %.2f) = %.4f\n", x, y, h);
        return h;
    }

    return it->second.getHeight(x, y);
}

// ---------------------------------------------------------------------------

float PathFinder::computeDistance(float x1, float y1, float x2, float y2)
{
    fprintf(stderr, "[pathfind] computeDistance (%.4f, %.4f) -> (%.4f, %.4f)\n", x1, y1, x2, y2);

    if (fabsf(x1) > WORLD_MAX || fabsf(y1) > WORLD_MAX ||
        fabsf(x2) > WORLD_MAX || fabsf(y2) > WORLD_MAX)
    {
        fprintf(stderr, "[pathfind] ERROR: coordinates exceed world bounds (±%.0f yards)\n", WORLD_MAX);
        return -1.0f;
    }

    if (!m_initialized)
    {
        fprintf(stderr, "[pathfind] ERROR: not initialized\n");
        return -1.0f;
    }

    ensureNavTilesLoaded(x1, y1, x2, y2);

    float z1 = getTerrainHeight(x1, y1);
    float z2 = getTerrainHeight(x2, y2);

    if (z1 == TERRAIN_INVALID_HEIGHT || z2 == TERRAIN_INVALID_HEIGHT)
    {
        fprintf(stderr, "[pathfind] ERROR: terrain height lookup failed (z1=%.2f z2=%.2f)\n", z1, z2);
        return -1.0f;
    }

    float startPos[3], endPos[3];
    toRecast(x1, y1, z1, startPos);
    toRecast(x2, y2, z2, endPos);
    fprintf(stderr, "[pathfind] recast start=(%.2f,%.2f,%.2f) end=(%.2f,%.2f,%.2f)\n",
            startPos[0], startPos[1], startPos[2],
            endPos[0],   endPos[1],   endPos[2]);

    const float extents[3] = { 2.0f, 4.0f, 2.0f };

    dtQueryFilter filter;
    dtPolyRef startRef = 0, endRef = 0;
    m_navQuery->findNearestPoly(startPos, extents, &filter, &startRef, nullptr);
    m_navQuery->findNearestPoly(endPos,   extents, &filter, &endRef,   nullptr);
    fprintf(stderr, "[pathfind] startRef=%llu endRef=%llu\n",
            (unsigned long long)startRef, (unsigned long long)endRef);

    if (!startRef || !endRef)
    {
        fprintf(stderr, "[pathfind] ERROR: could not find nearest poly (try widening extents?)\n");
        return -1.0f;
    }

    dtPolyRef polys[MAX_POLYS];
    int npolys = 0;
    m_navQuery->findPath(startRef, endRef, startPos, endPos,
                         &filter, polys, &npolys, MAX_POLYS);
    fprintf(stderr, "[pathfind] findPath: %d polys\n", npolys);

    if (!npolys)
        return -1.0f;

    // If the path doesn't reach the target polygon, clamp end to the last reachable poly.
    float endPosAdj[3];
    dtVcopy(endPosAdj, endPos);
    if (polys[npolys - 1] != endRef)
    {
        fprintf(stderr, "[pathfind] path incomplete — clamping to last reachable poly\n");
        m_navQuery->closestPointOnPoly(polys[npolys - 1], endPos, endPosAdj, nullptr);
    }

    float straightPath[MAX_POLYS * 3];
    unsigned char straightPathFlags[MAX_POLYS];
    dtPolyRef straightPathPolys[MAX_POLYS];
    int nstraight = 0;
    m_navQuery->findStraightPath(startPos, endPosAdj, polys, npolys,
                                 straightPath, straightPathFlags, straightPathPolys,
                                 &nstraight, MAX_POLYS);
    fprintf(stderr, "[pathfind] findStraightPath: %d points\n", nstraight);

    if (nstraight < 2)
        return -1.0f;

    float total = 0.0f;
    for (int i = 0; i < nstraight - 1; ++i)
    {
        const float* a = &straightPath[i * 3];
        const float* b = &straightPath[(i + 1) * 3];
        float dx = b[0] - a[0], dy = b[1] - a[1], dz = b[2] - a[2];
        total += sqrtf(dx*dx + dy*dy + dz*dz);
    }

    fprintf(stderr, "[pathfind] total distance: %.4f\n", total);
    return total;
}
